published/external/xdpapi.h

```diff
@@ -41,21 +49,92 @@ XDP_CREATE_PROGRAM_FN(
     );
 
 typedef
-HRESULT
+XDP_STATUS
 XDP_INTERFACE_OPEN_FN(
+    _In_ XDP_API_PROVIDER_BINDING_CONTEXT *ProviderBindingContext,
     _In_ UINT32 InterfaceIndex,
     _Out_ HANDLE *InterfaceHandle
     );
 
+#else
+
+typedef
+XDP_STATUS
+XDP_CREATE_PROGRAM_FN(
+    _In_ UINT32 InterfaceIndex,
+    _In_ const XDP_HOOK_ID *HookId,
+    _In_ UINT32 QueueId,
+    _In_ XDP_CREATE_PROGRAM_FLAGS Flags,
+    _In_reads_(RuleCount) const XDP_RULE *Rules,
+    _In_ UINT32 RuleCount,
+    _Out_ HANDLE *Program
+    );
+
+typedef
+XDP_STATUS
+XDP_INTERFACE_OPEN_FN(
+    _In_ UINT32 InterfaceIndex,
+    _Out_ HANDLE *InterfaceHandle
+    );
+
+#endif
+
 #include "afxdp.h"
 
-typedef struct _XDP_API_TABLE XDP_API_TABLE;
+typedef
+VOID *
+XDP_GET_ROUTINE_FN(
+    _In_z_ const CHAR *RoutineName
+    );
 
 //
 // The only API version currently supported.
 //
 #define XDP_API_VERSION_1 1
 
+#if defined(_KERNEL_MODE)
+
+typedef
+VOID
+XDP_DELETE_PROGRAM_FN(
+    _In_ HANDLE Program
+    );
+
+typedef
+VOID
+XDP_INTERFACE_CLOSE_FN(
+    _In_ HANDLE InterfaceHandle
+    );
+
+DEFINE_GUID(
+    NPI_XDPAPI_INTERFACE_ID,
+    0x1683b7b0, 0xcf44, 0x4757, 0x91, 0xf6, 0xc6, 0x19, 0x9b, 0x9d, 0x00, 0xfe);
+
+typedef struct _XDP_API_PROVIDER_DISPATCH {
+    XDP_GET_ROUTINE_FN *XdpGetRoutine;
+    XDP_CREATE_PROGRAM_FN *XdpCreateProgram;
+    XDP_DELETE_PROGRAM_FN *XdpDeleteProgram;
```

Comment:
i don't think we should have these - instead, let's have a common XdpCloseHandle routine, and plumb one more function pointer to XDP_FILE_OBJECT_HEADER.

Reply:
For me it is too low resolution comments. Not sure what are these for and how can be replaced.
@nigriMSFT could you help?

Reply:
@mtfriesen which functions would be replaced with this `XdpCloseHandle`?

Reply:
I got the meaning about XdpCloseHandle, but what do you mean about "XDP_FILE_OBJECT_HEADER". what feature do you want.

Reply:
The struct implementation of each one of these objects (program, interface, ..) has a XDP_FILE_OBJECT_HEADER member. Inside XDP_FILE_OBJECT_HEADER is a dispatch table. Currently the dispatch table only caters to the IRP patterns of user-mode. I think Michael is suggesting that we can refactor/extend this dispatch table such that it caters to both the user-mode and kernel mode APIs. The XdpCloseHandle entrypoints for each API should take the same core code path where the above mentioned dispatch table is used to call the object's Close/Delete routines. I might be getting some details wrong but that is my high level take

Reply:
As far as I checked, `XskCreate` (`XdpApiKernelXskCreate`) and `XdpInterfaceOpen` (`XdpApiKernelXdpInterfaceOpen`) set its `Header.ObjectType` to its types. However `XdpCreateProgram` (`XdpApiKernelXdpCreateProgram`) doesn't.
What @nigriMSFT said as **refactoring** means is XdpCreateProgram to set its `Header.ObjectType`?

Reply:
draft. XdpDeleteProgram is not unified as I commented.

https://github.com/microsoft/xdp-for-windows/pull/718